ballot/src/lib.rs

```diff
@@ -48,69 +87,120 @@ impl Ballot {
         Ok(())
     }
 
+    fn update_cache(&mut self, block_hash: &BlockHash, proof_type: BallotDrawResult) {
+        self.block_hash_cache
+            .lock()
+            .unwrap()
+            .put(*block_hash, proof_type);
+    }
+
+    fn query_cache(&self, block_hash: &BlockHash) -> Option<BallotDrawResult> {
+        self.block_hash_cache
+            .lock()
+            .unwrap()
+            .get(block_hash)
+            .cloned()
+    }
+
     /// Draw proof types based on the block hash.
-    pub fn draw(&self, block_hash: &BlockHash) -> Option<ProofType> {
+    pub fn draw(&mut self, block_hash: &BlockHash) -> BallotDrawResult {
+        if let Some(res) = self.query_cache(block_hash) {
```

Comment:
Cache operations that cross double lock access in concurrency may have data race issues.

Reply:
fixed